client-misk-jooq/src/test/kotlin/app/cash/backfila/client/misk/MiskJooqBackfillTests.kt

```diff
@@ -0,0 +1,497 @@
+      backfillRowKeys = WidgetCompoundKeyBackfillDbSetup.createNoMatching(transacter, clock, tokenGenerator),
+      backfila
+    )
+  ).map { backfillTestOption ->
+    DynamicTest.dynamicTest("no match test ${backfillTestOption.backfillClass.simpleName}") {
+      val run = backfillTestOption.createDryRun().apply { configureForTest() }
+      assertThat(run.partitionProgressSnapshot.values.single().utf8RangeStart()).isNotNull()
+      assertThat(run.partitionProgressSnapshot.values.single().utf8RangeEnd()).isNotNull()
+      assertThat(run.partitionProgressSnapshot.values.single().previousEndKey).isNull()
+
+      val scan1 = run.singleScan()
+      assertThat(scan1.batches).size().isEqualTo(1)
+      assertThat(scan1.batches.single().scanned_record_count).isEqualTo(5)
+      assertThat(scan1.batches.single().matching_record_count).isEqualTo(0)
+      app.cash.backfila.client.misk.testing.assertThat(run.partitionProgressSnapshot.values.single())
+        .isNotDone()
+      assertThat(run.partitionProgressSnapshot.values.single().previousEndKey).isNotNull
+
+      val scan2 = run.singleScan()
+      assertThat(scan2.batches).isEmpty()
+      app.cash.backfila.client.misk.testing.assertThat(run.partitionProgressSnapshot.values.single())
+        .isDone()
+      app.cash.backfila.client.misk.testing.assertThat(run).hasNoBatchesToRun().isComplete()
+    }
+  }
+
+  @TestFactory
+  fun withStartRange() = listOf(
+    JooqBackfillTestOptions(
+      backfillClass = JooqMenuTestBackfill::class,
+      backfillRowKeys = JooqMenuBackfillDbSetup.createSome(transacter),
+      backfila
+    ),
+    JooqBackfillTestOptions(
+      backfillClass = JooqWidgetCompoundKeyBackfill::class,
+      backfillRowKeys = WidgetCompoundKeyBackfillDbSetup.createSome(transacter, clock, tokenGenerator),
+      backfila
+    )
+  ).map { backfillOption ->
+    DynamicTest.dynamicTest("with start range test ${backfillOption.backfillClass.simpleName}") {
+      // Start at the 2nd id, so the 1st should be skipped.
+      val run = backfillOption.createDryRun(rangeStart = backfillOption.backfillRowKeys[1].toString())
+        .apply { configureForTest() }
+      assertThat(run.rangeStart).isEqualTo(backfillOption.backfillRowKeys[1].toString())
+      app.cash.backfila.client.misk.testing.assertThat(run.partitionProgressSnapshot.values.single())
+        .isNotDone()
+
+      run.singleScan()
+      assertThat(run.batchesToRunSnapshot.single().utf8RangeStart()).isEqualTo(
+        backfillOption.backfillRowKeys[1].toString()
+      )
+      assertThat(run.partitionProgressSnapshot.values.single().utf8PreviousEndKey()).isEqualTo(
+        backfillOption.backfillRowKeys[10].toString()
+      )
+
+      run.execute()
+      assertThat(run.backfill.idsRanDry).containsExactlyElementsOf(backfillOption.backfillRowKeys.slice(1..19))
+      assertThat(run.backfill.idsRanWet).isEmpty()
+    }
+  }
+
+  @TestFactory
+  fun withEndRange() = listOf(
+    JooqBackfillTestOptions(
```

Comment:
Is there a more efficient way of doing this?

Seems to be a lot of copypasta.

For Hibernate I created a base class and then had two implementations. There might be an even better way.

Reply:
I changed this to be a parameterized test. I preferred the more verbose version though as the test setup and the test were in the same place. But this not so bad either.